src/client.rs

```diff
@@ -269,38 +269,43 @@ impl NamedPipeClientReader {
 
     /// Read full message/bytes into a vec
     pub fn read_full(&self) -> Result<Vec<u8>, Error> {
-        let buffer_size = self.options.read_buffer_size;
-
-        let mut buffer: Vec<u8> = Vec::with_capacity(buffer_size as usize);
-
-        let mut buffer_to_read = buffer_size;
-        let mut buffer_ptr = 0;
-        let mut read_bytes = 0;
-
-        while unsafe {
-            ReadFileSys(
-                self.handle.0 .0,
-                buffer.as_mut_ptr().add(buffer_ptr) as *mut _,
-                buffer_to_read,
-                &mut read_bytes,
-                std::ptr::null_mut(),
-            ) == 0
-        } {
+        let buffer_size = std::cmp::max(self.options.read_buffer_size as usize, 1);
+
+        let mut buffer = Vec::new();
+        let mut total_read_bytes = 0;
+
+        loop {
+            let old_buffer_size = buffer.len();
+            let new_buffer_size = old_buffer_size + buffer_size;
+            buffer.resize(new_buffer_size, 0);
+
+            let buffer_end = &mut buffer[old_buffer_size..];
+            let mut read_bytes = 0;
+            let result = unsafe {
+                ReadFileSys(
+                    self.handle.0 .0,
+                    buffer_end.as_mut_ptr(),
+                    buffer_end.len() as u32,
+                    &mut read_bytes,
+                    std::ptr::null_mut(),
+                )
+            };
+
+            total_read_bytes += read_bytes as usize;
+
+            if result != 0 {
+                // Read was successful
+                buffer.resize(total_read_bytes, 0);
+                return Ok(buffer);
+            }
+
             let err = unsafe { GetLastError().unwrap_err() };
-            if err.code() != ERROR_MORE_DATA.into() {
+            if err.code() == ERROR_MORE_DATA.into() {
```

Comment:
I'm reluctant to add an empty `if` branch. I understand it's more for readability, though at this point it would be better to add a `continue` to it, but since it's already at the end, I think fallthrough makes more sense. We could still make it a little clearer like this, and leave the comment below so it's still clear

```rust
if err.code() != ERROR_MORE_DATA.into() {
    // An error occurred during reading
    return Err(err);
}

// Read succeeded, but this message has more data
```

Reply:
Ah oops, definitely didn't mean to leave the body actually empty! Definitely agree it's clearer with the way you suggested... I just updated the PR branch